### PR TITLE
 BAG dataset - calculate the northeast pixel corner rather than scaling the resolution, due to an incorrectly shifted northeast corner in some CARIS surveys

### PR DESCRIPTION
**Pull request by zacharyburnettNOAA**
## What does this PR do?

attempts to fix issue brought up in https://github.com/OSGeo/gdal/issues/1643

## What are related issues/pull requests?

https://github.com/OSGeo/gdal/issues/1643

### CODE REVIEW

File: gdal/frmts/hdf5/bagdataset.cpp [hunk 1]
```diff
@@ -3346,8 +3345,7 @@ void BAGDataset::LoadMetadata()
                 {
                     // Found with https://data.ngdc.noaa.gov/platforms/ocean/nos/coast/H12001-H14000/H12525/BAG/H12525_MB_4m_MLLW_1of2.bag
                     // https://github.com/OSGeo/gdal/issues/1643
-                    CPLDebug("BAG", "cornerPoints seems to use a PixelIsArea convention.");
-                    bPixelCenterConvention = false;
+                    CPLError(CE_Failure, CPLE_AppDefined,"Ignoring BAG Upper Right Corner.");
```
Comment (rouault) [inline gdal/frmts/hdf5/bagdataset.cpp]: I'd wish we maintain the ability to handle in a nice way those datasets, even if there are "illegal" given the specification (buggy products do exist in the wild, and when it is possible and reasonable code-wise to handle them, we should try). I'm OK with emitting a message, but rather a CE_Warning than a CE_Failure, and with a more explicit message saying that the product uses a PixelIsArea convention which is non-standard

File: gdal/frmts/hdf5/bagdataset.cpp [hunk 2]
```diff
@@ -4582,8 +4569,6 @@ void GDALRegister_BAG()
 "  <Option name='ZLEVEL' type='int' "
     "description='DEFLATE compression level 1-9' default='6' />"
 "  <Option name='BLOCK_SIZE' type='int' description='Chunk size' />"
-"  <Option name='CORNER_POINTS_EXTEND_HALF_PIXEL' type='boolean' description="
```
Comment (rouault) [inline gdal/frmts/hdf5/bagdataset.cpp]: I'm OK if we drop that capability, but then it must be undocumented as well in doc/source/drivers/raster/bag.rst

File: autotest/gdrivers/bag.py [hunk 3]
```diff
@@ -666,49 +665,29 @@ def test_bag_write_south_up():
...
-    gdal.Translate('/vsimem/out_non_standard.bag', 'data/byte.tif',
-                   creationOptions = ['VAR_DATE=2019-06-21',
-                                      'CORNER_POINTS_EXTEND_HALF_PIXEL=YES'])
-
-    f = gdal.VSIFOpenL('/vsimem/out.bag', 'rb')
-    assert f
-    data = gdal.VSIFReadL(1, 1000000, f)
-    gdal.VSIFCloseL(f)
-    gdal.Unlink('/vsimem/out.bag')
+    ds = gdal.Open('data/test_offset_ne_corner.bag')
```
Comment (rouault) [inline autotest/gdrivers/bag.py]: is that one completely inconsistent, or using a pixelisarea convention ? I'd wish we keep a test for a dataset with a pixelisarea convention

File: gdal/frmts/hdf5/bagdataset.cpp (hunk 1 above)
Comment (rouault) [reply gdal/frmts/hdf5/bagdataset.cpp]: In the final else case with the 3 CPLDebug() messages, you shoud also emit a CPLError(CE_Warning, ...) since that latter case is more annoying that the PixelIsArea case: here the georeferencing is clearly inconsistent !

File: gdal/frmts/hdf5/bagdataset.cpp [hunk 4]
```diff
@@ -3375,15 +3373,13 @@ void BAGDataset::LoadMetadata()
             }
 
             adfGeoTransform[0] = dfLLX;
-            adfGeoTransform[1] = (dfURX - dfLLX) / (m_nLowResWidth - (bPixelCenterConvention ? 1 : 0));
-            adfGeoTransform[3] = dfURY;
-            adfGeoTransform[5] = (dfLLY - dfURY) / (m_nLowResHeight - (bPixelCenterConvention ? 1 : 0));
+            adfGeoTransform[1] = dfResWidth;
```
Comment (rouault) [inline gdal/frmts/hdf5/bagdataset.cpp]: shoud be modiied to behave nicely when PixelIsArea convention is detected

File: gdal/frmts/hdf5/bagdataset.cpp (hunk 1 above)
Comment (GlenRice-NOAA) [reply gdal/frmts/hdf5/bagdataset.cpp]: Our understanding of the meaning of these errors is limited and we are more than happy to address these cases as you suggest.  Our only goal here is to ignore the upper right corner rather than scale the resolution.  Since BAGs should not be pixelisarea we wanted to remove that as an option for writing as well.

File: autotest/gdrivers/bag.py (hunk 3 above)
Comment (GlenRice-NOAA) [reply autotest/gdrivers/bag.py]: We were attempting to add a test case for BAGs that were being assessed as pixelisarea due to the upper right corner being offset to ensure they were being handled as intended.

File: gdal/frmts/hdf5/bagdataset.cpp [hunk 5]
```diff
@@ -3344,8 +3344,8 @@ void BAGDataset::LoadMetadata()
                     fabs((dfURY - dfLLY) / dfResHeight - m_nLowResHeight) < 1e-2 )
                 {
                     // Found with https://data.ngdc.noaa.gov/platforms/ocean/nos/coast/H12001-H14000/H12525/BAG/H12525_MB_4m_MLLW_1of2.bag
-                    // https://github.com/OSGeo/gdal/issues/1643
-                    CPLError(CE_Failure, CPLE_AppDefined,"Ignoring BAG Upper Right Corner.");
+                    // to address issue https://github.com/OSGeo/gdal/issues/1643
+                    CPLError(CE_Warning, CPLE_AppDefined, "cornerPoints not consistent with resolution given in metadata");
```
Comment (rouault) [inline gdal/frmts/hdf5/bagdataset.cpp]: I'd suggest a more specific message for that case: "cornerPoints and resolution in metadata seem to imply a PixelCorner convention, whereas BAG mandates PixelCenter. GDAL will georeference the product assuming this non-standard PixelCorner convention"

File: gdal/frmts/hdf5/bagdataset.cpp (hunk 4 above)
Comment (GlenRice-NOAA) [reply gdal/frmts/hdf5/bagdataset.cpp]: Our understanding of the previous code was that PixelIsArea was assumed when the upper left corner was not written correctly.  This was resulting in the resolution being scaled, which is not what was intended.  As was described in #1643, the BAG is defined by the lower lefter corner and the resolution.  The upper right corner is only written because of our understanding of  ISO 19115, and apparently is being written incorrectly because there is no check on its validity when written by the associated software.  We hope to correct this.

Comment (rouault) [reply gdal/frmts/hdf5/bagdataset.cpp]: > Our understanding of the previous code was that PixelIsArea was assumed when the upper left corner was not written correctly. This was resulting in the resolution being scaled, which is not what was intended.

your proposed changes regarding set adfGeoTransform[] look good, except than the half-pixel shift of the origin must not be done in the PixelIsArea/PixelCorner case.

Comment (GlenRice-NOAA) [reply gdal/frmts/hdf5/bagdataset.cpp]: We did the half shift to the extracted GeoTransform because it was consistent with the rest of the GDAL BAG driver.  It looked like the way the driver was written (and all the tests worked) took care of the shift during reading and writing, thus the GDAL BAG object took on a PixelIsCorner convention.  Should it not be that way?  Should we should modify the GeoTransform (and all the tests) to keep the BAG as PixelIsArea as a GDAL object? There must be an attribute we would set that informs a transform to or from other object type of any conversion that needs to be made to the associated GeoTransform.

Comment (rouault) [reply gdal/frmts/hdf5/bagdataset.cpp]: > We did the half shift to the extracted GeoTransform because it was consistent with the rest of the GDAL BAG driver

That's fine. I meant that when the convention of the product is detected to be the non-standard/erroneous PixelCorner/PixelIsArea then this half-pixel shift is not needed, since it is already in the georeferencing convention used by GDAL.

Comment (GlenRice-NOAA) [reply gdal/frmts/hdf5/bagdataset.cpp]: From our perspective the non-standard detection indicates that the upper right corner was written incorrectly and thus a warning should be thrown.  It does not mean the BAG is PixelIsCorner.  Since the code (now) always recalculates the upper right corner there is no need to make the half pixel shift conditional.  The resolution and lower left corner are always correct and the raster is always PixelIsArea.

Comment (rouault) [reply gdal/frmts/hdf5/bagdataset.cpp]: >  It does not mean the BAG is PixelIsCorner

Hum, perhaps... Is there a way of checking what the correct georeferencing would be for a product like
https://data.ngdc.noaa.gov/platforms/ocean/nos/coast/H12001-H14000/H12525/BAG/H12525_MB_4m_MLLW_1of2.bag which was referenced in #1643 ? If indeed this is just an issue with the computation of the upper-right corner, then https://github.com/OSGeo/gdal/pull/1728/commits/d455285c787046d05e1215496441f7248bc4094a should be reverted. This is becoming confusing... :-)

**Comment by zacharyburnettNOAA** [comment]
No, I don't think there are any more issues. Thank you for reviewing the changes.

File: gdal/frmts/hdf5/bagdataset.cpp (hunk 4 above)
Comment (GlenRice-NOAA) [reply gdal/frmts/hdf5/bagdataset.cpp]: We went back and compared the source raster with the derived BAG for a few of our products and found the lower left corner and resolution were correct.  The upper right corner was reported as being one grid resolution away from what is reported in the BAG, so clearly it was being ignored by the software.  We think this confirms the steps taken here.
We often say these things are not hard, but there are confusing.  Thanks for accepting the work.
